### PR TITLE
Refactor diagnose agent

### PR DESCRIPTION
**Pull request by samplise**
### What changes were proposed in this pull request?

Refactor the implementation of diagnose agent on worker.

### Why are the changes needed?

We have a new design for diagnose system.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

File: dlrover/python/elastic_agent/torch/training.py [hunk 1]
```diff
@@ -642,16 +648,16 @@ def _invoke_run(self, role: str = DEFAULT_ROLE) -> RunResult:
                 return run_result
             elif state in {WorkerState.UNHEALTHY, WorkerState.FAILED}:
                 logger.error(f"The worker fails with {run_result.failures}")
-                self._report_failure_to_master(run_result.failures)
```
Comment (liyzcj) [inline dlrover/python/elastic_agent/torch/training.py:645]: we need a  faults detection here，not just depend on training process status.

File: dlrover/python/diagnose/common/inference_chain.py [hunk 2]
```diff
@@ -69,6 +74,15 @@ def same_inference(inference1: Inference, inference2: Inference) -> bool:
         return False
 
 
+def include_inference(infs: List[Inference], inf: Inference) -> bool:
```
Comment (BalaBalaYi) [inline dlrover/python/diagnose/common/inference_chain.py:77]: ```is_inference_included```

File: dlrover/python/diagnose/common/inference_chain.py [hunk 3]
```diff
@@ -69,6 +74,15 @@ def same_inference(inference1: Inference, inference2: Inference) -> bool:
...
+    if not infs or not inf:
+        return False
+    for i in infs:
+        if same_inference(i, inf):
```
Comment (BalaBalaYi) [inline dlrover/python/diagnose/common/inference_chain.py:81]: ```is_same_inference```

File: dlrover/python/diagnose/inferencechain/inferenceoperator/check_failure_node_operator.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    InferenceDescription,
+    InferenceName,
+    InferenceOperator,
+)
+from dlrover.python.diagnose.datacollector.training_log_collector import (
+    TrainingLogCollector,
+)
+
+
+class CheckFailureNodeOperator(InferenceOperator):
```
Comment (BalaBalaYi) [inline dlrover/python/diagnose/inferencechain/inferenceoperator/check_failure_node_operator.py:30]: Need annotations.

File: dlrover/python/diagnose/datacollector/training_log_collector.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dlrover.python.diagnose.common.diagnose_data import TrainingLog
+from dlrover.python.diagnose.datacollector.data_collector import DataCollector
+from dlrover.python.util.file_util import read_last_n_lines
+
+
+class TrainingLogCollector(DataCollector):
```
Comment (BalaBalaYi) [inline dlrover/python/diagnose/datacollector/training_log_collector.py:19]: Need annotations.

File: dlrover/python/diagnose/inferencechain/inferenceoperator/check_failure_node_operator.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+            return True
+        else:
+            return False
+
+    def infer(self, inferences: List[Inference]) -> List[Inference]:
+        if (
+            len(inferences) == 0
+            or not inferences[0].configs
+            or InferenceConfigKey.LOG_FILE not in inferences[0].configs
+            or InferenceConfigKey.ERRORS not in inferences[0].configs
```
Comment (BalaBalaYi) [inline dlrover/python/diagnose/inferencechain/inferenceoperator/check_failure_node_operator.py:49]: We can't implement like 'inferences[0]' here. Need to provide a function to search specified inference by name.

File: dlrover/python/diagnose/inferencechain/inferenceoperator/check_failure_node_operator.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+            ]
+        log_file = inferences[0].configs[InferenceConfigKey.LOG_FILE]
+        errors = inferences[0].configs[InferenceConfigKey.ERRORS]
+        error_codes = errors.split("#")
+
+        collector = TrainingLogCollector(log_file, 5000)
+        training_log = collector.collect_data()
+        logs = training_log.logs
+        if not logs or len(logs) == 0:
+            logger.error(f"fail to collect training log from {log_file}")
```
Comment (BalaBalaYi) [inline dlrover/python/diagnose/inferencechain/inferenceoperator/check_failure_node_operator.py:66]: Could it also be a normal situation if the log is empty? Shouldn't it be appropriate to print an error here?

File: dlrover/python/elastic_agent/diagnose/diagnose_agent.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    InferenceName,
+    include_inference,
+)
+from dlrover.python.diagnose.inferencechain.inference_chain import (
+    InferenceChain,
+)
+from dlrover.python.elastic_agent.master_client import MasterClient
+
+
+class DiagnoseAgent:
```
Comment (BalaBalaYi) [inline dlrover/python/elastic_agent/diagnose/diagnose_agent.py:41]: Need annotations.

File: dlrover/python/elastic_agent/torch/training.py [hunk 9]
```diff
@@ -379,6 +381,8 @@ def __init__(
         start_method="spawn",
         exit_barrier_timeout: float = 300,
         log_dir: Optional[str] = None,
+        training_log_file: str = "",
```
Comment (BalaBalaYi) [inline dlrover/python/elastic_agent/torch/training.py:384]: Considering such variables will be added in the future, it is necessary to define a unified POJO to encapsulate these variables.

File: dlrover/python/elastic_agent/torch/training.py [hunk 10]
```diff
@@ -405,6 +409,9 @@ def __init__(
         self._save_ckpt_executor = ThreadPoolExecutor(max_workers=1)
         self._save_ckpt_future = None
         self._log_file = os.getenv(NodeEnv.TRAINING_LOG_FILE, "")
+        self._diagnose_agent = DiagnoseAgent(
```
Comment (BalaBalaYi) [inline dlrover/python/elastic_agent/torch/training.py:412]: "error code is 507035" is internal logic, should not appear in the input parameters.

File: dlrover/python/elastic_agent/torch/training.py [hunk 11]
```diff
@@ -642,21 +649,17 @@ def _invoke_run(self, role: str = DEFAULT_ROLE) -> RunResult:
...
-                        f"{self._remaining_failovers}/{spec.max_restarts}"
-                        f" attempts left; will restart worker group"
-                    )
-                    self._remaining_failovers -= 1
-                    self._restart_workers(self._worker_group)
-                else:
-                    self._stop_workers(self._worker_group)
-                    self._worker_group.state = WorkerState.FAILED
+
+                worker_context = WorkerContext(
```
Comment (BalaBalaYi) [inline dlrover/python/elastic_agent/torch/training.py:654]: Need more test cases for coverage for this part involves core logic for fault tolerance.

File: dlrover/python/master/diagnosis/diagnosis.py [hunk 12]
```diff
@@ -65,12 +64,4 @@ def stop(self):
...
-        logger.info("Start to diagnose failures")
-        while True:
-            observed_problems = self.diagnostician.observe_training()
-            for problem in observed_problems:
-                logger.info(f"observed problems: {problem}")
-                root_causes = self.diagnostician.diagnose_failure(problem)
-                for root_cause in root_causes:
-                    logger.info(f"identify root cause: {root_cause}")
-            time.sleep(180)
+        pass
```
Comment (BalaBalaYi) [inline dlrover/python/master/diagnosis/diagnosis.py:67]: This is the core implementation section and needs to be fully implemented.

File: dlrover/python/diagnose/inferencechain/inferenceoperator/check_training_hang_operator.py [hunk 13]
```diff
@@ -24,8 +23,8 @@
 
 
 class CheckTrainingHangOperator(InferenceOperator):
```
Comment (BalaBalaYi) [inline dlrover/python/diagnose/inferencechain/inferenceoperator/check_training_hang_operator.py:25]: These modifications need to be refactored after this PR: https://github.com/intelligent-machine-learning/dlrover/pull/1228.

File: dlrover/python/diagnose/common/diagnose_data.py [hunk 14]
```diff
@@ -13,8 +13,7 @@
 
 from abc import ABCMeta, abstractmethod
 from datetime import datetime
-
-from dlrover.python.util.file_util import read_last_n_lines
+from typing import List, Optional
```
Comment (BalaBalaYi) [inline dlrover/python/diagnose/common/diagnose_data.py:16]: ```diagnose_data.py``` should be named as ```diagnosis_data.py```

File: dlrover/python/elastic_agent/diagnose/diagnose_agent.py [hunk 15]
```diff
@@ -0,0 +1,99 @@
+# Copyright 2024 The DLRover Authors. All rights reserved.
```
Comment (BalaBalaYi) [inline dlrover/python/elastic_agent/diagnose/diagnose_agent.py:1]: ```diagnose_agent.py``` should be named as ```diagnosis_agent.py```

File: dlrover/python/diagnose/common/diagnose_data.py (hunk 14 above)
Comment (BalaBalaYi) [reply dlrover/python/diagnose/common/diagnose_data.py:16]: Also, there are 2 'diagnose_data.py' in common and master package. Need to unified.

**Review by BalaBalaYi — CHANGES_REQUESTED** [review]

File: dlrover/python/elastic_agent/torch/training.py (hunk 1 above)
Comment (samplise) [reply dlrover/python/elastic_agent/torch/training.py:645]: We plan to add this in future pr.

File: dlrover/python/master/diagnosis/diagnosis.py (hunk 12 above)
Comment (samplise) [reply dlrover/python/master/diagnosis/diagnosis.py]: This pr focuses on worker side. Master side will be refactored in the next pr.

**Review by BalaBalaYi — APPROVED** [review]
LGTM

**Review by BalaBalaYi — APPROVED** [review]
lgtm

**Pull request merged**